aten/src/ATen/native/cuda/DilatedMaxPool2d.cu

```diff
@@ -57,15 +62,91 @@ __global__ void MaxPoolForward(const int nthreads, const scalar_t* bottom_data,
   }
 }
 
-static const int BACKWARD_THREADS = 256;
+template <typename scalar_t, typename accscalar_t>
+C10_LAUNCH_BOUNDS_1(CUDA_MAX_THREADS)
+__global__ void MaxPoolForwardNHWC(const int nthreads, const scalar_t* bottom_data,
+                                   const int num, const int channels, const int height,
+                                   const int width, const int pooled_height, const int pooled_width,
+                                   const int kernel_h, const int kernel_w, const int stride_h,
+                                   const int stride_w, const int pad_h, const int pad_w,
+                                   const int dilation_h, const int dilation_w,
+                                   const int in_stride_c, const int in_stride_h, const int in_stride_w,
+                                   scalar_t* top_data, int64_t* top_mask) {
+
+  extern __shared__ int smem[];
+  scalar_t *out_cached = reinterpret_cast<scalar_t*>(smem);
+
+  // flattening cta for pre-computation & smem initialization;
+  int thread_id = threadIdx.x + blockDim.x * (threadIdx.y + blockDim.y * threadIdx.z); //globalId
+  int block_size = blockDim.x * blockDim.y * blockDim.z; //block size
+
+  // use shared memory to store temporary output value. This is simply to
+  // reduce register usage.
+  for (int i = thread_id; i < channels*blockDim.y*blockDim.z; i+= block_size) {
+    out_cached[i] = scalar_t(0.0);
+    out_cached[2 * i + 1] = scalar_t(0.0);
+  }
+
+  __syncthreads();
+
+  top_data = top_data + blockIdx.x * pooled_height * pooled_width * channels;
+  top_mask = top_mask + blockIdx.x * pooled_height * pooled_width * channels;
+  bottom_data = bottom_data + blockIdx.x * channels * height * width;
+
+  out_cached = &out_cached[2 * (threadIdx.z * blockDim.y + threadIdx.y) * channels];
+
+  int oH = (pooled_height + gridDim.z-1) / gridDim.z;
+  int oW = (pooled_width + gridDim.y-1) / gridDim.y;
+  int ostartH = threadIdx.z + blockIdx.z*oH;
+  int oendH = ::min(ostartH+oH, pooled_height);
+  int ostartW = threadIdx.y + blockIdx.y*oW;
+  int oendW = ::min(ostartW+oW, pooled_width);
+
+  for (int oh = ostartH; oh < oendH; oh+=blockDim.z) {
+    for (int ow = ostartW; ow < oendW; ow+=blockDim.y) {
+      int hstart = oh * stride_h - pad_h;
+      int wstart = ow * stride_w - pad_w;
+      int hend = min(hstart + (kernel_h - 1) * dilation_h + 1, height);
+      int wend = min(wstart + (kernel_w - 1) * dilation_w + 1, width);
+      while(hstart < 0)
+        hstart += dilation_h;
+      while(wstart < 0)
+        wstart += dilation_w;
+      for (int ih = hstart; ih < hend; ih++) {
+        for (int iw = wstart; iw < wend; iw++) {
+          const scalar_t *ptr_input = bottom_data + ih * in_stride_h + iw * in_stride_w;
+          for(int c = threadIdx.x; c < channels; c+= blockDim.x) {
+            scalar_t val = ptr_input[c];
+            scalar_t maxval = out_cached[2 * c];
```

Comment:
NIT: `maxval` doesn't seem to be necessary here;

Reply:
Fixed